Fix text shadow cutoff on android

insacc opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:
-Changes the shadow style span from a CharacterStyle to a `ReplacementSpan` in order the fix the issue on android where the shadow is cut off even with added padding as the characterStyle doesnt account for the shadow.

This accounts for the shadow but also moves the text back to its original position so on the UI there wouldn't be any shifts when the shadow is applied.

- Updates the gradient span to make sure that it matches the ios implementation

- Updates the strokeWidth for the text stroke span to make sure that it matches the ios implementation

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->


## Test Plan:
- Tested locally on android with different states to make sure that it works as expected
<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

| Before | After | 
| --- | --- | 
| <img width="1043" height="472" alt="cut-off-v1" src="https://github.com/user-attachments/assets/e31fc950-40d1-43e2-b5aa-b48728f82d31" />|<img width="1038" height="383" alt="cut-off-fixed-v1" src="https://github.com/user-attachments/assets/41e07597-ce4c-44fd-bedd-6c3ca94e274f" />|

insacc committed:
Fix the shadow cut off on android and update gradient

Attemp to fix text shadow cutoff

Update the text shadow on android to prevent cutoff

Expand the shadow on if padding is provided

Add the missing import

Add some logging

Make shadow span use padding

Attempt to fix text shadow cutoff on android

Another attempt at fixing the text shadow cutoff

Another attempt at fixing the text shadow cutoff issue on Android.

Add more logging

Add more logs

Add more and more logs

Implement shadow offset compensation in ReactTextView

Make sure to keep the text position stable when the shadow is added.

Update linear gradient span to match iOS behavior.

Address gradient issues

Another attempt to fix the text stroke width

Revert some of the gradient changes

Update shader mode

Update the implementation once more

Revert to width expansion approach - shadow cutoff confirmed without it

Add more logging to the ReactTextView

Remove debug logging

Remove remaining debug logs from text rendering files

Remove unused padding parameters from ShadowStyleSpan

- Remove padding constructor parameters that were never used in getSize() or draw()
- Remove updatePadding() method that was never called
- Remove padding retrieval code from ReactBaseTextShadowNode
- Both old and new architectures now consistently use 4-parameter constructor

Align new arch (Fabric) shadow handling with old arch (Paper)

- Remove shadowTopOffset from PreparedLayoutTextView
- Vertical shadow space is already handled via font metrics adjustment
- Both architectures now only compensate horizontally
- Matches old arch behavior and comment: 'vertical doesn't need compensation'

Remove unused getShadowDy() method and min import

- getShadowDy() was only used for vertical compensation which we removed
- kotlin.math.min import is not used anywhere in the file

insacc committed:
Remove no longer needed changes

insacc committed:
Remove the unnecessary clipping changes

Flewp reviewed (commented):
Unless we plan on upstreaming this functionality to react-native, I would suggest putting as much of this feature code as possible internally. For example, making `ShadowStyleSpan` an open class and implementing our own internally, or providing callbacks to shim logic in the fork where the actual business logic exists internally.

These kinds of changes, especially feature + followup style, cause non-trivial overhead when we're trying to bump React Native versions and reapply every single change on top of the new version. They also destabilize features like this over time - if upstream makes API changes or migrates files from Java to Kotlin (I see ReactTextView is still Java), these merges can be error prone. 

All being said, if you need to merge as is, go for it. But it would be valuable to refactor to minimize our changes to react native as a fast follow.

insacc committed:
Decouple discord shadow style span from react native

insacc committed:
Use the custom discord shadow style span in the text layout manager

insacc committed:
Remove unnecessary import

insacc committed:
Fixes the api visibility

insacc committed:
Add more public visibility to the discord shadow style span

insacc committed:
Remove unused getShadowDy() and getShadowColor() from DiscordShadowStyleSpan

These getters are not used anywhere in the codebase.
Only getShadowRadius() and getShadowDx() are used by getShadowAdjustment().

insacc committed:
Fix truncation issues with text shadows
